04 Research Environment/03 Datasets/08 Equity Options/10 Get Price Model Data.html

```diff
@@ -1,61 +1,166 @@
+var vega = new Vega(optionContract, mirrorOption: mirrorContract, optionModel: OptionPricingModelType.ForwardTree);
+var theta = new Theta(optionContract, mirrorOption: mirrorContract, optionModel: OptionPricingModelType.ForwardTree);
+var rho = new Rho(optionContract, mirrorOption: mirrorContract, optionModel: OptionPricingModelType.ForwardTree);</pre>
+    </div>
+
     <li>Iterate through the historical data and calculate the values.</li>
-     <div class="section-example-container">
+    <div class="section-example-container">
         <pre class="python">df = pd.DataFrame()
-for slice in history:
-    underlying_price = None
-    underlying_volatility = None
-
-    # Update the security with QuoteBar information
-    for bar in slice.QuoteBars.Values:
-        qb.Securities[bar.Symbol].SetMarketPrice(bar)
-
-    # Update the security with TradeBar information
-    for bar in slice.Bars.Values:
-        symbol = bar.Symbol
-        security = qb.Securities[symbol]
-        security.SetMarketPrice(bar)
-
-        if security.Type == SecurityType.Equity:
-            underlying_volatility = security.VolatilityModel.Volatility
-            underlying_price = security.Price
-            continue
-        
-        # Create the Option contract
-        contract = OptionContract.Create(symbol.Underlying, bar.EndTime, security, underlying_price)
-        contract.LastPrice = bar.Close
-
-        # Evaluate the price model to get the IV, Greeks, and theoretical price
-        result = security.PriceModel.Evaluate(security, None, contract)
-        greeks = result.Greeks
-        
-        # Append the data to the DataFrame
-        data = {
-            "IV" : result.ImpliedVolatility,
-            "Delta": greeks.Delta,
-            "Gamma": greeks.Gamma,
-            "Vega": greeks.Vega,
-            "Rho": greeks.Rho,
-            "Theta": greeks.Theta,
-            "LastPrice": contract.LastPrice,
-            "Close": security.Close,
-            "theoreticalPrice" : result.TheoreticalPrice,
-            "underlyingPrice": underlying_price,
-            "underlyingVolatility": underlying_volatility
-        }
-        right = "Put" if symbol.ID.OptionRight == 1 else "Call"
-        index = pd.MultiIndex.from_tuples([(symbol.ID.Date, symbol.ID.StrikePrice, right, symbol.Value, bar.EndTime)], names=["expiry", "strike", "type", "symbol", "endTime"])
-        df = pd.concat([df, pd.DataFrame(data, index=index)])</pre>
+
+for bar in history:
+    # Update the option indicators with indicator data points of the 3 symbols
+    equity_point = IndicatorDataPoint(equity_symbol, bar[equity_symbol].EndTime, bar[equity_symbol].Close)
+    option_point = IndicatorDataPoint(option_contract, bar[option_contract].EndTime, bar[option_contract].Close)
+    mirror_point = IndicatorDataPoint(mirror_contract, bar[mirror_contract].EndTime, bar[mirror_contract].Close)
+
+    iv.Update(equity_point)
+    iv.Update(option_point)
+    iv.Update(mirror_point)
```

Comment:
Could we use the helper methods? If we update the security, it will update the consolidators and updates all indicators.

Reply:
I followed other research docs page on using manual indicators. Auto-updating might not be working since we're in research environment?